### PR TITLE
change: enable PR build

### PR DESCRIPTION
**Pull request by ChuyangDeng**
*Issue #, if available:*

*Description of changes:*
- add buildspec.yml to enable PR build
- project is already added in cdk package, will run webhub script after buildspec is merged

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: buildspec.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      # run unit tests
+      - tox -e py27,py36 test/unit
+
+      # create pip archive (tar name will be something like sagemaker_mxnet_serving_container-1.0.0.tar.gz, but the Dockerfiles expect sagemaker_mxnet_serving_container.tar.gz)
+      - python3 setup.py sdist
+      - mv dist/sagemaker_mxnet_serving_container-*.tar.gz dist/sagemaker_mxnet_serving_container.tar.gz
+      - cp dist/sagemaker_mxnet_serving_container.tar.gz docker/$FRAMEWORK_FULL_VERSION/py3/sagemaker_mxnet_serving_container.tar.gz
+
+      # build images
+      - python3 scripts/build_all.py --version $FRAMEWORK_FULL_VERSION --account $ACCOUNT --repo $ECR_REPO
```
Comment (laurenyu) [inline buildspec.yml]: the other PR builds add a build ID to the tag - e.g. https://github.com/aws/sagemaker-mxnet-container/blob/master/buildspec.yml#L47

File: buildspec.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
+version: 0.2
+
+env:
+  variables:
+    FRAMEWORK_FULL_VERSION: '1.4.1'
+    FRAMEWORK_SHORT_VERSION: '1.4'
+    AWS_DEFAULT_REGION: 'us-west-2'
+    ECR_REPO: 'sagemaker-mxnet-serving'
```
Comment (laurenyu) [inline buildspec.yml]: the PR builds use a test repo: https://github.com/aws/sagemaker-mxnet-container/blob/master/buildspec.yml#L10

File: buildspec.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      - cmd="IGNORE_COVERAGE=- tox -e py36 -- test/integration/local --processor gpu --py-version 2,3 --framework-version $FRAMEWORK_FULL_VERSION --region $AWS_DEFAULT_REGION --docker-base-name $ecr_image"
+      - remote-test --test-cmd "$cmd" --github-repo $GITHUB_REPO --branch master --setup-file $SETUP_FILE
+
+      # run sagemaker integration tests
+      - IGNORE_COVERAGE=- tox -e py36 -- test/integration/sagemaker -n 16 --py-version 2,3 --processor cpu,gpu --region $AWS_DEFAULT_REGION --docker-base-name $ECR_REPO --aws-id $ACCOUNT --framework-version $FRAMEWORK_FULL_VERSION --reruns 3 --reruns-delay 5
+
+      # run ei tests
+      - IGNORE_COVERAGE=- tox -e py36 -- test/integration/sagemaker/test_elastic_inference.py -n 2 --py-version 2,3 --processor cpu --accelerator-type $EI_ACCELERATOR_TYPE --region $AWS_DEFAULT_REGION --docker-base-name "$ECR_REPO-eia" --aws-id $ACCOUNT --framework-version $FRAMEWORK_FULL_VERSION --reruns 3
+
+    finally:
```
Comment (laurenyu) [inline buildspec.yml]: also might want to delete the images: https://github.com/aws/sagemaker-mxnet-container/blob/master/buildspec.yml#L110

File: buildspec.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      - create-key-pair
+      - launch-ec2-instance --instance-type $GPU_INSTANCE_TYPE --ami-name dlami-ubuntu
+
+      # run gpu local integration tests
+      - printf "$SETUP_CMDS" > $SETUP_FILE
+      - ecr_image="$ACCOUNT.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$ECR_REPO"
+      - cmd="IGNORE_COVERAGE=- tox -e py36 -- test/integration/local --processor gpu --py-version 2,3 --framework-version $FRAMEWORK_FULL_VERSION --region $AWS_DEFAULT_REGION --docker-base-name $ecr_image"
+      - remote-test --test-cmd "$cmd" --github-repo $GITHUB_REPO --branch master --setup-file $SETUP_FILE
+
+      # run sagemaker integration tests
```
Comment (laurenyu) [inline buildspec.yml]: it'll be helpful to not run these when there are only doc changes, etc. - see https://github.com/aws/sagemaker-mxnet-container/blob/master/buildspec.yml#L73

File: buildspec.yml (hunk 1 above)
Comment (ChuyangDeng) [reply buildspec.yml]: attached tag with build ID.

File: buildspec.yml (hunk 2 above)
Comment (ChuyangDeng) [reply buildspec.yml]: Using sagemaker-test repo now.

File: buildspec.yml (hunk 3 above)
Comment (ChuyangDeng) [reply buildspec.yml]: Cleaned up test images in the end

File: buildspec.yml (hunk 4 above)
Comment (ChuyangDeng) [reply buildspec.yml]: skipping tests when there's only doc change

File: buildspec.yml [hunk 5]
```diff
@@ -3,61 +3,106 @@ version: 0.2
...
 
-      # run cpu local integration tests
-      - IGNORE_COVERAGE=- tox -e py36 -- test/integration/local --py-version 2,3 --processor cpu --framework-version $FRAMEWORK_FULL_VERSION --region $AWS_DEFAULT_REGION --docker-base-name $ECR_REPO
+      # build cpu image
+      - cpu_dockerfile="Dockerfile.cpu"
+      - CPU_TAG="$FRAMEWORK_FULL_VERSION-cpu-py$CPU_PY_VERSION-$build_id"
 
-      # push docker images to ECR
-      - python3 scripts/publish_all.py --version $FRAMEWORK_FULL_VERSION --account $ACCOUNT --repo $ECR_REPO
+      - docker build -f $build_dir/py$CPU_PY_VERSION/$cpu_dockerfile --build-arg framework_support_installable=$tar_name -t $PREPROD_IMAGE:$CPU_TAG .
```
Comment (laurenyu) [inline buildspec.yml]: I think it'd be better to use the same building scripts and test commands as the release build, and just modify `build_all.py` and `publish_all.py` to take the build ID as an argument, and then add something like:

```python
if args.build_id:
    tag += '-{}'.format(args.build_id)
```

Comment (ChuyangDeng) [reply buildspec.yml]: Added `--build-id` option in build_all.py and publish_all.py

**Pull request merged**